modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,330 @@
+            self.list_of_partitions_to_combine = drained
+            self.call_queue = []
+
+        PandasOnRayDataframeVirtualPartition.drain_call_queue = dr
+
+    def add_query(
+        self,
+        func: Callable,
+        is_output: bool = False,
+        repartition_after: bool = False,
+        fan_out: bool = False,
+        pass_partition_id: bool = False,
+        reduce_fn: Optional[Callable] = None,
+        output_id: Optional[int] = None,
+    ):
+        """
+        Adds a query to the current pipeline.
+
+        Parameters
+        ----------
+        func : Callable
+            DataFrame query to perform.
+        is_output : bool, default: False
+            Whether this query should be designated as an output query. If `True`, the output of
+            this query is passed both to the next query and directly to postprocessing.
+        repartition_after : bool, default: False
+            Whether the dataframe should be repartitioned after this query. Currently,
+            repartitioning is only supported if there is 1 partition prior.
+        fan_out : bool, default: False
+            Whether to fan out this node. If True and only 1 partition is passed as input, the
+            partition is replicated `num_partition` times, and the function is called on each.
+            The `reduce_fn` must also be specified.
+        pass_partition_id : bool, default: False
+            Whether to pass the numerical partition id to the query.
+        reduce_fn : Callable
+            The reduce function to apply if `fan_out` is set to True. This takes the `num_partition`
+            partitions that result from this query, and combines them into 1 partition.
+        output_id : int, default None
+            An id to assign to this node if it is an output.
+        """
+        self.nodes_list.append(
+            PandasQuery(
+                func,
+                is_output,
+                repartition_after,
+                fan_out,
+                pass_partition_id,
+                reduce_fn,
+            )
+        )
+        if is_output:
+            self.outputs.append(self.nodes_list[-1])
+            if output_id:
+                if self.node_to_id is None:
+                    self.node_to_id = {}
+                self.node_to_id[self.outputs[-1]] = output_id
+            curr_node = self.outputs[-1]
+            curr_node.operators = self.nodes_list[:-1]
+            self.nodes_list = []
+
+    def _complete_nodes(self, list_of_nodes, ptns):
+        """
+        Run a sub-query end to end.
+
+        Parameters
+        ----------
+        list_of_nodes: list of PandasQuery
+            The functions that compose this query.
+        ptns: list of PandasOnRayDataframeVirtualPartition
+            The ptns that compose the dataframe that is input to this sub-query.
+        """
+        for node in list_of_nodes:
+            if node.fan_out:
+                ptns[0] = ptns[0].force_materialization()
```

Comment:
Fan out and repartition are currently only supported if we only have 1 partition. Here, I've implemented them to ignore the other partitions if there are more than 1. Should we keep this approach or should we be more defensive (i.e. error out if there's more than 1)?

Reply:
We now warn here.